### ISSUE TITLE
Race condition when trying to update criteria 

### ISSUE THREAD

**Issue by Trinovantes** [SanaMobile/sana.protocol_builder#318]
If I try to update a criteria in 2 quick requests, there's a race condition that causes the first response to return 2 nodes and second response to return 0 nodes.

Request 1:
```
{
   "page":28,
   "id":60,
   "last_modified":"2016-02-21T23:26:14.286201Z",
   "created":"2016-02-21T23:26:14.252864Z",
   "conditions":[
      {
         "criteria_element":null,
         "node_type":"EQUALS",
         "value":"asd"
      }
   ]
}
```

Request 2:
```
{
   "page":28,
   "id":60,
   "last_modified":"2016-02-21T23:26:14.286201Z",
   "created":"2016-02-21T23:26:14.252864Z",
   "conditions":[
      {
         "criteria_element":null,
         "node_type":"EQUALS",
         "value":"asdf"
      }
   ]
}
```

Response 1:
```
{
   "id":60,
   "page":28,
   "last_modified":"2016-02-21T23:26:17.789627Z",
   "created":"2016-02-21T23:26:14.252864Z",
   "conditions":[
      {
         "id":172,
         "parent":null,
         "criteria_element":null,
         "node_type":"EQUALS",
         "value":"asd",
         "last_modified":"2016-02-21T23:26:17.778803Z",
         "created":"2016-02-21T23:26:17.778831Z",
         "show_if":60,
         "children":[

         ]
      },
      {
         "id":173,
         "parent":null,
         "criteria_element":null,
         "node_type":"EQUALS",
         "value":"asdf",
         "last_modified":"2016-02-21T23:26:17.779719Z",
         "created":"2016-02-21T23:26:17.779743Z",
         "show_if":60,
         "children":[

         ]
      }
   ]
}
```

Response 2:
```
{
   "id":60,
   "page":28,
   "last_modified":"2016-02-21T23:26:17.753356Z",
   "created":"2016-02-21T23:26:14.252864Z",
   "conditions":[

   ]
}
```

**Comment by Kapin** [comment]
Looking into this. However, from a client-perspective, wouldn't it make sense to ensure that the first save went through before the second one?

**Comment by Trinovantes** [comment]
I worked around this by adding a sleep between requests. I think it's safer to make sure the server can handle these cases than trusting the client. (e.g. maybe an MITM can perform a replay attack to make the user lose data?)

**Comment by Kapin** [comment]
@Trinovantes I'll have it throttle then. However, I'm implementing debouncing so this isn't dumb.

**Comment by Trinovantes** [comment]
You can't make the entire request atomic? Throttling on the server assumes the database operation will finish before the next request comes in

**Comment by Trinovantes** [comment]
Also occasionally get this 500 server error when trying to make a lot of changes

```
IntegrityError at /api/conditionals/155
insert or update on table "api_conditionnode" violates foreign key constraint "api_conditio_parent_id_2c82b1aa73da1709_fk_api_conditionnode_id"
DETAIL:  Key (parent_id)=(3336) is not present in table "api_conditionnode".


Request Method: PUT
Request URL: http://localhost:8000/api/conditionals/155
Django Version: 1.8.7
Python Executable: /usr/bin/python
Python Version: 2.7.6
Python Path: ['/home/stephen/Documents/sanaprotocolbuilder.me/src-django', '/usr/lib/python2.7', '/usr/lib/python2.7/plat-x86_64-linux-gnu', '/usr/lib/python2.7/lib-tk', '/usr/lib/python2.7/lib-old', '/usr/lib/python2.7/lib-dynload', '/usr/local/lib/python2.7/dist-packages', '/usr/lib/python2.7/dist-packages', '/usr/lib/python2.7/dist-packages/PILcompat', '/usr/lib/python2.7/dist-packages/gtk-2.0', '/usr/lib/pymodules/python2.7', '/usr/lib/python2.7/dist-packages/ubuntu-sso-client', '/home/stephen/Documents/sanaprotocolbuilder.me/src-django']
Server time: Wed, 24 Feb 2016 05:48:34 +0000
Installed Applications:
('django.contrib.admin',
 'django.contrib.sessions',
 'django.contrib.contenttypes',
 'django.contrib.auth',
 'django.contrib.staticfiles',
 'rest_framework',
 'rest_framework.authtoken',
 'django_nose',
 'corsheaders',
 'django_extensions',
 'api',
 'authentication',
 'mailer')
Installed Middleware:
('django.contrib.sessions.middleware.SessionMiddleware',
 'corsheaders.middleware.CorsMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'sanaprotocolbuilder.middleware.ExceptionLoggingMiddleware')

Traceback:
File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py" in get_response
  132.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/usr/local/lib/python2.7/dist-packages/django/views/decorators/csrf.py" in wrapped_view
  58.         return view_func(*args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/rest_framework/viewsets.py" in view
  87.             return self.dispatch(request, *args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py" in dispatch
  466.             response = self.handle_exception(exc)
File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py" in dispatch
  463.             response = handler(request, *args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/rest_framework/mixins.py" in update
  70.         self.perform_update(serializer)
File "/usr/local/lib/python2.7/dist-packages/rest_framework/mixins.py" in perform_update
  74.         serializer.save()
File "/usr/local/lib/python2.7/dist-packages/rest_framework/serializers.py" in save
  186.             self.instance = self.update(self.instance, validated_data)
File "/home/stephen/Documents/sanaprotocolbuilder.me/src-django/api/serializer.py" in update
  185.                 self.__create_nested_children(condition_data, instance)
File "/home/stephen/Documents/sanaprotocolbuilder.me/src-django/api/serializer.py" in __create_nested_children
  195.                     self.__create_nested_children(child_data, None)
File "/home/stephen/Documents/sanaprotocolbuilder.me/src-django/api/serializer.py" in __create_nested_children
  195.                     self.__create_nested_children(child_data, None)
File "/home/stephen/Documents/sanaprotocolbuilder.me/src-django/api/serializer.py" in __create_nested_children
  197.                 models.ConditionNode.objects.create(show_if=show_if, **data)
File "/usr/local/lib/python2.7/dist-packages/django/db/models/manager.py" in manager_method
  127.                 return getattr(self.get_queryset(), name)(*args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/django/db/models/query.py" in create
  348.         obj.save(force_insert=True, using=self.db)
File "/home/stephen/Documents/sanaprotocolbuilder.me/src-django/api/models.py" in save
  186.         super(ConditionNode, self).save()
File "/usr/local/lib/python2.7/dist-packages/django/db/models/base.py" in save
  734.                        force_update=force_update, update_fields=update_fields)
File "/usr/local/lib/python2.7/dist-packages/django/db/models/base.py" in save_base
  762.             updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
File "/usr/local/lib/python2.7/dist-packages/django/db/transaction.py" in __exit__
  225.                         connection.commit()
File "/usr/local/lib/python2.7/dist-packages/django/db/backends/base/base.py" in commit
  173.         self._commit()
File "/usr/local/lib/python2.7/dist-packages/django/db/backends/base/base.py" in _commit
  142.                 return self.connection.commit()
File "/usr/local/lib/python2.7/dist-packages/django/db/utils.py" in __exit__
  98.                 six.reraise(dj_exc_type, dj_exc_value, traceback)
File "/usr/local/lib/python2.7/dist-packages/django/db/backends/base/base.py" in _commit
  142.                 return self.connection.commit()

Exception Type: IntegrityError at /api/conditionals/155
Exception Value: insert or update on table "api_conditionnode" violates foreign key constraint "api_conditio_parent_id_2c82b1aa73da1709_fk_api_conditionnode_id"
DETAIL:  Key (parent_id)=(3336) is not present in table "api_conditionnode".
```

**Comment by Trinovantes** [comment]
It also just occurred to me that if we're destroying and rebuilding the tree every time, why not just store a JSON blob in the ShowIf model? Then we'd need to parse the JSON blob (either formally with into actual Python objects or informally by doing recursive tree traversal) to be able to perform semantic analysis/verification on it.

This also solves the potential problem of overflowing the primary key of Conditional Nodes in production since I'm already going through thousands just testing.

**Comment by Kapin** [comment]
@Trinovantes that was my original design. I'm perfectly ok going back to that because it would be drastically simpler.

**Comment by Kapin** [comment]
Fixed by #349
